lib/recursive-methods.rb

```diff
@@ -1,49 +1,106 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 0
+  raise ArgumentError, "must be a number greater than 0." if n < 0
+
+  return 1 if n == 1
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(1)
+
+def reverse(string, rev = "", first = 0)
+  if rev.length == string.length
+    return rev
+  end
+  rev = string[first] + rev
+  reverse(string, rev, first + 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(1)
+
+#All the variables that you have to change would go in the parameters.
+def reverse_inplace(s, first = 0, last = s.length - 1)
+  #This is usually the condition in the loop, reversed. Example while first < last becuase your saying when should I stop not when should i keep going.
+  if first >= last
+    return s
+  end
+  #this is the inside of the loop
+  temp = s[first]
+  s[first] = s[last]
+  s[last] = temp
+
+  # this is the bottom of the loop where you say i += 1
+  return reverse_inplace(s, first + 1, last - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(1)
+# Space complexity: O(1)
+
+#I want to add n to itself,but I don't now what my base case would be
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return ears = (n + n)
+  bunny(n)
```

Comment:
Just for readability

```suggestion
  return ears + bunny(n)
```